### PR TITLE
Feature/dsp 280 component updates

### PR DESCRIPTION
**Pull request by steveatinsolvency**
Initial work to reduce the components down to use a single controller.

### CODE REVIEW

File: src/Application/INSS.Platform.Portal.Application/Services/FormService.cs [hunk 1]
```diff
@@ -1,76 +1,128 @@
...
-        form.PopAllNavigationHistory();
-        await _formStateService.SaveAsync(_userSessionResolver.GetUserId(), form);
+        if (model is ConfirmModel confirm)
+        {
+            return await GetPostComfirmPageUrlAsync(form, confirm);
+        }
+
+        BaseModel currentModel = form.FindPage(model.PageUrl);
+
+        if (currentModel is SectionModel section)
```
Comment (davecook-ins) [inline src/Application/INSS.Platform.Portal.Application/Services/FormService.cs]: Could use a switch statement here to be explicit in what should happen if currentModel is SummaryListModel (even if that is nothing)
e.g.
```
switch (currentModel)
{
    case SectionModel section:
        section.IsComplete = true;
        break;
    case SummaryListModel:
        break;
    default:
        model.CopyTo(currentModel);
        break;
}
```

File: src/Application/INSS.Platform.Portal.Application/Services/FormService.cs [hunk 2]
```diff
@@ -1,76 +1,128 @@
...
+        return page is SectionModel ? page.PageUrl + "/summary" : page.PageUrl;
     }
 
-    private async Task<FormModel> CreateFormModelAsync()
+    public async Task<string> ChangeAsync(string itemId)
     {
-        string sessionId = _userSessionResolver.GetUserId();
+        FormModel form = await _formStateService.GetAsync();
+        SummaryListModel summaryList = form.FindSummaryList(itemId);
+        BaseModel page = summaryList.Items.First(i => i.Id == itemId);
```
Comment (davecook-ins) [inline src/Application/INSS.Platform.Portal.Application/Services/FormService.cs]: Can summaryList.Items contain multiple items with the same Id?  If not use Single() rather than First()

File: src/Application/INSS.Platform.Portal.Application/Validation/DefaultModelStateValidator.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿using INSS.Platform.Portal.Domain;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace INSS.Platform.Portal.Application.Validation;
+
+public sealed class DefaultModelStateValidator : IModelStateValidator<BaseModel>
+{
+    public Task ValidateAsync(ModelStateDictionary modelState, BaseModel model)
+    {
+        // Do nothing
```
Comment (davecook-ins) [inline src/Application/INSS.Platform.Portal.Application/Validation/DefaultModelStateValidator.cs]: Comment not required.

File: src/Domain/INSS.Platform.Portal.Domain/BaseModel.cs [hunk 4]
```diff
@@ -1,10 +1,113 @@
...
+        BindingFlags.SetProperty;
+    
+    protected BaseModel()
+    {
+        Id = Guid.NewGuid().ToString("D");
+        Kind = GetType().Name;
+        ViewName = $"_{GetType().Name.Replace("Model", "")}";
+    }
+    
+    public string Id { get; set; }
```
Comment (davecook-ins) [inline src/Domain/INSS.Platform.Portal.Domain/BaseModel.cs:25]: Inconsistency in setting default values.

File: src/Domain/INSS.Platform.Portal.Domain/BaseModel.cs [hunk 5]
```diff
@@ -1,10 +1,113 @@
...
+        {
+            property.SetValue(clone, property.GetValue(this));
+        }
+        
+        return clone;
+    }
+    
+    public void Reset()
+    {
+        Id = Guid.NewGuid().ToString("D");
```
Comment (davecook-ins) [inline src/Domain/INSS.Platform.Portal.Domain/BaseModel.cs:66]: Not that bothered about this but the "D" format is the default format, there is no benefit to using .ToString("D") over .ToString()

File: src/Presentation/INSS.Platform.Portal.Web.Components/Extensions/FormCollectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            
+            result[entry.Key] = ChangeType(entry.Value.ToString(), property.PropertyType);
+        }
+
+        return result;
+    }
+    
+    private static object? ChangeType(object? value, Type conversionType)
+    {
+        // Handle nullable types
```
Comment (davecook-ins) [inline src/Presentation/INSS.Platform.Portal.Web.Components/Extensions/FormCollectionExtensions.cs]: Not sure you need these comments.

File: src/Presentation/INSS.Platform.Portal.Web.Components/Extensions/FormCollectionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        Type targetType = Nullable.GetUnderlyingType(conversionType) ?? conversionType;
+
+        // Null handling for nullable types
+        if (value is null || value == DBNull.Value)
```
Comment (davecook-ins) [inline src/Presentation/INSS.Platform.Portal.Web.Components/Extensions/FormCollectionExtensions.cs]: Is DBNull.Value ever a possibility? As we are unlikely to be using ADO.Net or taking values directly from a DB surely you only need to test for null.

File: src/Presentation/INSS.Platform.Portal.Web.Components/Views/Shared/_Address.cshtml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+﻿@addTagHelper GovUk.Frontend.AspNetCore.TagHelpers.*, GovUk.Frontend.AspNetCore
```
Comment (davecook-ins) [inline src/Presentation/INSS.Platform.Portal.Web.Components/Views/Shared/_Address.cshtml]: Can the @addTagHelper in all of these views be added into _ViewImports ?

File: src/Presentation/INSS.Platform.Portal.Web/Factories/WebAppFormModelFactory.cs [hunk 9]
```diff
@@ -1,40 +1,60 @@
...
-        AddressModel currentAddress = new() { Title = "Address" };
-        sectionAboutYou.AddPage(currentAddress);
-        SummaryListModel summaryList =new() { Title = "Your Addresses"};
-        sectionAboutYou.AddPage(summaryList);
-        ConfirmModel confirm = new() { Title = "Confirm Address Deletion" }; 
-        sectionAboutYou.AddPage(confirm);
-
-        form.Initialize();
-
         /*
```
Comment (davecook-ins) [inline src/Presentation/INSS.Platform.Portal.Web/Factories/WebAppFormModelFactory.cs]: Not sure the comment example is required.  Any examples if needed could be added to the method documentation block.

File: test/Application/INSS.Platform.Portal.Application.Tests/Services/FormSerializationServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    [Fact]
+    public void FormWithModels_SerializeAndDeserialize_ChecksFormSerialization()
+    {
+        _mockModelTypeService.Setup(m => m.GetModelTypes()).Returns([typeof(AddressModel), typeof(BankAccountModel)]);
+        FormModel form = CreateFormModel();
+
+        string json = _service.Serialize(form);
+        FormModel deserializedForm = _service.Deserialize(json);
+
+        Assert.Equal(form.Id, deserializedForm.Id);
```
Comment (davecook-ins) [inline test/Application/INSS.Platform.Portal.Application.Tests/Services/FormSerializationServiceTests.cs:28]: AssertionScope would be nicer for all of these tests, so that we can easily identify all assertion failures rather than just the first.

**Review by davecook-ins — CHANGES_REQUESTED** [review]
Looks good, probably not a job for now, but going forward I think we should document all interfaces/classes/methods etc, even it is is just for the purpose of intellisense.

File: src/Application/INSS.Platform.Portal.Application/Services/FormService.cs (hunk 1 above)
Comment (steveatinsolvency) [reply src/Application/INSS.Platform.Portal.Application/Services/FormService.cs]: Will add a switch with a comment for the summary list

File: src/Application/INSS.Platform.Portal.Application/Services/FormService.cs (hunk 2 above)
Comment (steveatinsolvency) [reply src/Application/INSS.Platform.Portal.Application/Services/FormService.cs]: It is probably a bit moot but single will have to continue scanning the list for other entries to ensure it is the only one. First just gets a the first item it finds. Marginal performance gain here (nanoseconds) but as a general principal it isn't always sensible to assume single is the best option.

File: src/Domain/INSS.Platform.Portal.Domain/BaseModel.cs (hunk 4 above)
Comment (steveatinsolvency) [reply src/Domain/INSS.Platform.Portal.Domain/BaseModel.cs:25]: Assume this is PathName and PreviousPageUrl - moved to ctor

File: src/Presentation/INSS.Platform.Portal.Web.Components/Extensions/FormCollectionExtensions.cs (hunk 7 above)
Comment (steveatinsolvency) [reply src/Presentation/INSS.Platform.Portal.Web.Components/Extensions/FormCollectionExtensions.cs]: AI generated. Will remove

File: src/Presentation/INSS.Platform.Portal.Web.Components/Views/Shared/_Address.cshtml (hunk 8 above)
Comment (steveatinsolvency) [reply src/Presentation/INSS.Platform.Portal.Web.Components/Views/Shared/_Address.cshtml]: Moved

File: src/Presentation/INSS.Platform.Portal.Web/Factories/WebAppFormModelFactory.cs (hunk 9 above)
Comment (steveatinsolvency) [reply src/Presentation/INSS.Platform.Portal.Web/Factories/WebAppFormModelFactory.cs]: Removed

File: test/Application/INSS.Platform.Portal.Application.Tests/Services/FormSerializationServiceTests.cs (hunk 10 above)
Comment (steveatinsolvency) [reply test/Application/INSS.Platform.Portal.Application.Tests/Services/FormSerializationServiceTests.cs:28]: I am going to leave this for now as FluentAssertions has a paid model and I am not sure if we would be in breach of its non-commercial license. I see it is scattered across your tests:

Key Points of the New License

Non-Commercial Use: The new license allows for modification and redistribution of FluentAssertions as long as it is for non-commercial purposes. This means that individuals or organizations using the library for personal projects, educational purposes, or other non-commercial activities can continue to use it without any licensing fees.

Commercial Use: For commercial entities, the new license requires a paid subscription. This includes any use of FluentAssertions in commercial products or services. The definition of commercial use is broad and includes any activities by a commercial entity, even if the specific use of the library does not directly generate revenue.

Contributions: Contributions to the FluentAssertions project are still allowed, but contributors must consent to their changes becoming part of the new license terms and waive their rights. This ensures that any modifications or enhancements made by the community can be incorporated into future versions of the library under the new licensing model.

I think this one might need to defer to others as to whether we should be using it. I will reach out to Gary in the first instance and get a steer on the Insolvency Service take on it. 

Leave you stuff in for now but don't be surprised if we have to remove it

File: src/Domain/INSS.Platform.Portal.Domain/BaseModel.cs (hunk 4 above)
Comment (davecook-ins) [reply src/Domain/INSS.Platform.Portal.Domain/BaseModel.cs:25]: Sorry, I was referring to the assignment of string.Empty to only some of the string props.

I think I need to do a training course on GitHub reviews :) surely it can't be that bad a tool that you cannot highlight specific lines of code, surely it must be an operator issue.

File: src/Domain/INSS.Platform.Portal.Domain/BaseModel.cs (hunk 5 above)
Comment (steveatinsolvency) [reply src/Domain/INSS.Platform.Portal.Domain/BaseModel.cs:66]: This will be replaced by a primitive to represent the Id.

**Review by davecook-ins — APPROVED** [review]
looks good
